Adds installing cells as part of mech construction and fixes some mech construction bugs.

GLACoding opened the pull request:
[Changelogs]: #

:cl: GLA Coding
tweak: Cells must now be installed in mechs when being constructed, this step is always before the application of internal armor.
fix: Combat mechs now get stats upgrades from their scanning modules and capacitors
/:cl:

- Mechs must now have a cell inserted during construction (keeps it consistent with borg construction and means roboticists dont need to go through maintanance protocols to change a cell on a mech they just made.

- Combat mechs now use checkparts() and apply the upgrades from their capacitors and scanning modules properly (the spawn_result() in construction paths was not being called at all so ugrades were not applied).

- Reworked the spawn_result code in construction paths so that it is now used, blackbox will now correctly log the construction of mechs.

GLACoding committed:
spawn_mecha_result() code

Function that places the cell from construction from the holder into the new mech, and check and applies scanning module/capacitor upgrades for combat mechs

GLACoding committed:
New holder icons

No change made to the new sprites as of current.

GLACoding committed:
New ripley construction code

GLACoding committed:
New gygax construction code

GLACoding committed:
New fireripley construction code

GLACoding committed:
New honker construction code

GLACoding committed:
New durand construction code

GLACoding committed:
New phazon construction code

GLACoding committed:
New odysseus construction code

GunHog commented:
Emperor bless you!  I have long wanted to see this change happen!

MimicFaux commented:
yaaaasssss

MoreRobustThanYou commented:
Finally i'll have less people asking over mentorhelp on hippie "how to replace cell in mech"

AnonymousNow commented:
Remind me, what's the standard battery for mechs before this change? 10k, wasn't it? Because it's always been worth swapping them out for a highcap, at least.

Cyberboss commented on code/datums/helper_datums/construction_datum.dm:
```diff
@@ -62,6 +62,17 @@
+		var/obj/item/c = locate(/obj/item/stock_parts/cell) in holder
+		c.forceMove(m)
+		m.cell = c
+		if(findtext(result,"/obj/mecha/combat/"))
```
result needs to be stringified first, I believe

GLACoding replied:
result is already a string

Cyberboss replied:
There's a difference between a type and string. This looks like a type because of [this line](https://github.com/tgstation/tgstation/pull/30942/files#diff-b1dba44e1d25a22a29321794be968b19R67). In fact, the check should be `ispath(result, /obj/mecha/combat)`

Cyberboss commented on code/datums/helper_datums/construction_datum.dm:
```diff
@@ -62,6 +62,17 @@
+		if(findtext(result,"/obj/mecha/combat/"))
+			m.CheckParts(holder.contents)
+		qdel(holder)
+	return
```
remove trailing return

Cyberboss commented on code/game/mecha/mecha_construction_paths.dm:
```diff
@@ -196,86 +208,105 @@
 			else
+				user.visible_message("[user] prys the power cell from [holder].", "<span class='notice'>You pry the power cell from [holder].</span>")
+				var/obj/item/I = locate(/obj/item/stock_parts/cell) in holder
+				I.loc = get_turf(holder)
```
I.forceMove(holder.drop_location())

Cyberboss commented on code/game/mecha/mecha_construction_paths.dm:
```diff
@@ -513,57 +555,74 @@
 			else
+				user.visible_message("[user] prys the power cell from [holder].", "<span class='notice'>You pry the power cell from [holder].</span>")
+				var/obj/item/I = locate(/obj/item/stock_parts/cell) in holder
+				I.loc = get_turf(holder)
```
ditto

Cyberboss commented on code/game/mecha/mecha_construction_paths.dm:
```diff
@@ -695,96 +764,114 @@
-				MS.amount = 5
+				user.visible_message("[user] prys the power cell from [holder].", "<span class='notice'>You pry the power cell from [holder].</span>")
+				var/obj/item/I = locate(/obj/item/stock_parts/cell) in holder
+				I.loc = get_turf(holder)
```
ditto

Cyberboss commented on code/game/mecha/mecha_construction_paths.dm:
```diff
@@ -1098,57 +1207,74 @@
 			else
+				user.visible_message("[user] prys the power cell from [holder].", "<span class='notice'>You pry the power cell from [holder].</span>")
+				var/obj/item/I = locate(/obj/item/stock_parts/cell) in holder
+				I.loc = get_turf(holder)
```
ditto

Cyberboss commented on code/game/mecha/mecha_construction_paths.dm:
```diff
@@ -1403,84 +1539,101 @@
 			else
+				user.visible_message("[user] prys the power cell from [holder].", "<span class='notice'>You pry the power cell from [holder].</span>")
+				var/obj/item/I = locate(/obj/item/stock_parts/cell) in holder
+				I.loc = get_turf(holder)
```
ditto

Cyberboss commented on code/game/mecha/mecha_construction_paths.dm:
```diff
@@ -1609,88 +1773,106 @@
 			else
+				user.visible_message("[user] prys the power cell from [holder].", "<span class='notice'>You pry the power cell from [holder].</span>")
+				var/obj/item/I = locate(/obj/item/stock_parts/cell) in holder
+				I.loc = get_turf(holder)
```
how many times have i made this comment? God mech code is so shitty

GunHog commented:
https://github.com/tgstation/tgstation/blob/96718e9648935eae26cba8b89727751c2ddf4aec/code/game/mecha/mecha.dm#L219

Mechs start with 15K cells. That is, normal cells var-edited to be 15K. I do not know why this is needed when we could just start with them high-caps...

GLACoding commented:
Yep as gunhog said. Oddly enough the dark gygax spawns with a 30K cell, but the dark mauler does not.
For now admin/event spawned mechs will have the regular 15K cell.

AnonymousNow commented:
Ah, that's what it is - they come out looking like normal batteries, so the massive energy draw seems consistent.

Awareness of batteries in mechs should hopefully encourage roboticists to think about interchanging them. Ta much.

GLACoding committed:
Cyberboss Requested Fixes #1

Cyberboss commented on code/datums/helper_datums/construction_datum.dm:
```diff
@@ -62,6 +62,16 @@
+		m.cell = c
+		if(findtext(result,"/obj/mecha/combat/"))
+			m.CheckParts(holder.contents)
+		qdel(holder)
```
QDEL_NULL(holder)

GLACoding committed:
Cyberboss Requested Fixes #2

Turns result in mecha_construction_paths to types, also removes result in phazon_chassis as it has no purpose.
changes qdel to QDEL_NULL

GLACoding committed:
Removes old cell from add_cell() in the Initialize()

AnturK commented on code/datums/helper_datums/construction_datum.dm:
```diff
@@ -62,6 +62,18 @@
+		c.forceMove(m)
+		m.cell = c
+		if(ispath(result, /obj/mecha/combat))
+			m.CheckParts(holder.contents)
```
Do you actually need this typecheck ? CheckParts is obj/ function isn't it. And shouldn't cell updates be in there.

GLACoding replied:
CheckParts calls /obj/mecha/combat/CheckParts(list/parts_list) so only effects combat mechs. However i could move this to /obj/mecha and combine all the checks.

KorPhaeron commented:
Fantastic PR

GLACoding committed:
Move CheckParts to obj/mecha and move cell check to CheckParts
